ldc1612: handle i2c errors

nefelim4ag opened the pull request:
I feel somewhat obligated to finish i2c error handling, or at least push it a little forward after https://github.com/Klipper3d/klipper/pull/6689

I got some idea of how error handling happens around accelerometers, and can incorporate it into other i2c bulk data sensors.

So, here is a suggestion for LDC1612, the host code already handles cases where there are errors bit set in MSB. Then it should be easy to unconditionally set any of them, if i2c got an error, because `ret` is more than 0 on error, and less than 8.
```
    def _convert_samples(self, samples):
        ...
        for ptime, val in samples:
            mv = val & 0x0fffffff
            if mv != val:
                self.last_error_count += 1
```
If it would be better, we can manually hardcode enum values for consistency, to avoid assumptions.

`query_status` handling is done in the same way as in the `sensor_adxl345`.
Technically `> 64` would be enough (`1 << 6 == 64`).
![image](https://github.com/user-attachments/assets/0ba9a37b-0e3a-4e7e-a371-eb9d631e09b0)

Not tested, because I just don't have ldc1612.

Thanks.

ZeSlammy commented:
So ELI5, would this work on Temp, pressure, Gaz sensors plugged onto a PCB for a Stealth Max filter ?
I somehow have NACK i2C errors and they are seen as critical even in Kalico/DangerKlipper

nefelim4ag commented:
> So ELI5

This PR will touch only inductive bad sensors, like BTT Eddy.
So, not yet.
I hope, there will be a series of patch sets with error handling in different places, where it is easy to do with the current code infrastructure. But that will not touch Temp, Pressure, or Gaz sensors.

To generally handle I2C errors, as discussed in the previous topic,
there should be some command changes, and forwarding error information back to the host where they will be handled (possibly retried once). This can break something, this is a lot of work and will take several iterations.

> Temp, pressure, Gaz sensors plugged onto a PCB for a Stealth Max filter ?
> I somehow have NACK i2C errors and they are seen as critical even in Kalico/DangerKlipper

I don't know anything about your setup and think you can try to ask for help on the discourse about what you can do.
If you have separate MCUs for sensors and they do not control heaters.
I can only suggest something like:
```
 void i2c_shutdown_on_err(int ret)
 {
+    return;
     switch (ret) {
     case I2C_BUS_NACK:
         shutdown("I2C NACK");
```
To disable I2C errors completely. This is bad, this is a hack and if you don't know how to do it, you should not.

Hope it helps.

thijstriemstra commented:
> I feel somewhat obligated to finish i2c error handling

Thanks!

KevinOConnor commented:
Thanks for working on this.  I'll try to test this on my ldc1612.  Alas, I'm quite behind on projects so it may take me some time.

Cheers,
-Kevin

Supergun777 commented:
Patch applied, everything works as before, don't know how to reproduce I2C errors

nefelim4ag commented:
Just rebased, I did recheck the code if there is a appropriate error handling on in the LDC1612 code on the MCU and the Host side.

Seems it also should correctly report "Eddy current sensor error" on I2C communication error during the homing procedure.
```
    // High 4 bits are for errors
    d[0] |= ret << 4;

    // Check for endstop trigger
    uint32_t data =   ((uint32_t)d[0] << 24)
                    | ((uint32_t)d[1] << 16)
                    | ((uint32_t)d[2] << 8)
                    | ((uint32_t)d[3]);
    check_home(ld, data);
 ....
check_home(struct ldc1612 *ld, uint32_t data)
    if (data > 0x0fffffff) {
        // Sensor reports an issue - cancel homing
        ld->homing_flags = 0;
```

Thanks.

nefelim4ag commented:
I spent some more time with sensors and all of the application notes.
I guess we can utilize the under/over range bits: https://www.ti.com/lit/an/snoa959/snoa959.pdf

Under-range can't happen because we do not use offset, and if we do want it can be done in software to compact the bulk data output, and/or pack more status bits.

_My (carto) board would output something about raw values in the range: 3_000_000...4_000_000. BTT Eddy, AFAIK, uses 12MHz, and has a similar frequency range. So the raw values would be twice as much. 6_000_000..8_000_000._

Overrange, on the other hand, is always `0xfffffff`.

Watchdog/Zero count is something that should not happen, and honestly, both of them sound fatal to me.

nefelim4ag commented:
Patch works.
```
LDC1612 starting 'carto' measurements
LDC1612: I2C IO error
LDC1612: I2C IO error
```

_BTW, watchdog errors can be produced by shortening the coil._

Thanks.

nefelim4ag commented:
Okay, I've refactored it a little.
Added some docs which should describe errors for people who are not me.
It is honestly hard to check and reproduce everything.

Verified to happen:
- Firmware watchdog - with no clock signal or with low ODR.
- Over range - reducing clock input closer to sensor frequency (24MHz -> 4MHz in my case).
- Conversion watchdog - by closing the coil pins
- Zero count - by closing the coil pins
- Over range - by closing the coil pins
- I2C IO error - by closing of the I2C pins.

Messages in the log looks like so:
```
ldc1612: Amplitude Low/High (24)
ldc1612: Amplitude Low/High (36)
ldc1612: Amplitude Low/High (24)
ldc1612: Frequency over valid range (3)
ldc1612: Conversion Watchdog timeout (1)
ldc1612: Zero conversion count (1)
ldc1612: Zero conversion count (1)
ldc1612: Frequency over valid range (3)
ldc1612: Zero conversion count (1)
ldc1612: Zero conversion count (1)
```

To reduce amplite warnings rate, we can technically disable the amplitude high bit, I think.
But for now messages are still here.

That is it on my side.

Thanks,
-Timofey

KevinOConnor commented on klippy/extras/ldc1612.py:
```diff
@@ -161,12 +161,25 @@ def clear_home(self):
+                flags = val >> 28
+                if (flags & (0x8 | 0x4)) == (0x8 | 0x4):
+                    log_once("I2C IO error")
+                    continue
```
I wonder if we could improve this error reporting interface so that it is more "regular".  Perhaps we could specify that if the top 15 bits are all ones (`if mv >> 16 == 0xffff`) it indicates a specific error condition code is in the low bits (`error_code = mv & 0xffff`).  And then use the enumeration system to lookup a string for the error (`error_enums = mcu.get_enumerations("ldc1612_sensor_errors:")` ... `error_msg = error_enums.get(error_code); log_once(error_msg)`).

It is possible that the chip could send a message with all high bits set.  In that case, the mcu could detect that condition and report an error for that (`if data >> 16 == 0xffff { error_code = MY_DATA_ERROR; }` ... DECL_ENUMERATION("ldc1612_sensor_errors:", "sensor data invalid" MY_DATA_ERROR);` ).

Normally mcu enumerations describe parameters in host/mcu messages, but we could use a special character (eg, ":") to denote an enumeration not intended to directly describe a message parameter.

KevinOConnor replied:
> Probably the middle ground is to still send the bits as before, but in case of I2C errors do 0xffff encoding like above?

Right.  I don't think we want to reencode the error bits already present in the `REG_DATA0` register - as there's no confusion on the meaning of those bits.  I think it would be preferable to have a more regularized encoding for errors not already in `REG_DATA0` - for example: invalid data (all high bits set), i2c error, etc.

Using enumerations is just a "random idea" - I don't have strong feelings about it one way or another.

nefelim4ag commented on src/sensor_ldc1612.c:
```diff
@@ -44,6 +45,39 @@ check_intb_asserted(struct ldc1612 *ld)
 
+// Default TRSYNC_TIMEOUT, limits ODR > 40
+#define WATCHDOG_TIMEOUT timer_from_us(25000)
+#define WATCHDOG_TIMEOUT_ERROR (1<<4)
```
Sure.
If you don't mind.

I would suggest to have at least some host side temporary duct tape, where `EddyGatherSamples` would wait for samples to be recieved at least once (even if, it will be done not in this PR).
It is surprisingly easy to make sensor hung (at least for me with static_pwm_clock or when messing up with registers).

nefelim4ag commented:
Thanks, I hope I've addressed everything.

-Timofey

nefelim4ag commented:
Looks good to me overall.

Ah, yeah, I just was not really successful with filtering I2C error data from the stream, because the only solution that I obviously see is to copy data into the new array, and so, I decided it is "rare situation, whatever".
Yes, I think this sort of host dropping the I2C error data is the right approach.

I agree with the general intention to separate code paths, because otherwise it is quite a mess here.
I just wanted to avoid code duplication. Also, I like the separete handling of I2C status/data messages, and detection of 0xffff.

I will try to incorporate your work, thanks!

-Timofey.

KevinOConnor commented:
> This is my attempt to ensure generous sanity check in case if coil is disconnected
Or frequency is misconfigured

> But there should be safe guard for the tap probe, which only cares about diffs of the value.

I understand and I agree.  It's a hard coded range limit in the mcu code though.  I think a value passed from the host would be preferable, even if that limit is hard coded in the host.

In any case, it is not a big deal either way.  If you are ready, I'll commit this PR.

Cheers,
-Kevin

nefelim4ag commented:
Appreciated,
Let's commit it then.

-Timofey

_Now all LDC based probes should be fully practically functional._
